### PR TITLE
QEP 2: Enhanced Array Literals

### LINKED ISSUE #573: Type Argument Inference Fails if the same Type Parametrized Callable is used Multiple Times
This code highlights an issue with type argument inference:
```
operation` Foo<'T> (arg : 'T) : Unit { }
 
operation DoTwo<'A, 'B>(op1 : ('A => Unit), arg1 : 'A, op2: ('B => Unit), arg2 : 'B) : Unit {
    op1(arg1);
    op2(arg2);
}
 
operation CombineTwo<'A, 'B>(op1 : ('A => Unit), op2: ('B => Unit)) : (('A, 'B) => Unit) {
    return DoTwo(op1, _, op2, _);
}
 
operation Main() : Unit {
    (CombineTwo(Foo, Foo))("Hi", 3);
}
```
The expression in the Main operation has an issue here. The type parameter resolution data structure in the syntax tree doesn't explicitly contain any way of keeping track of which instance of a type parameter the resolution is for. It only contains the callable name, the type parameter name, and the resolving type. In the code above, in Main, the expression has two conflicting resolutions because it is trying to resolve `Foo.T` to `String` and its trying to resolve `Foo.T` to `Int`. It sees this conflict because it can't differentiate between the resolution for the first Foo and the resolution for the second Foo. As a result it throws an error complaining about ambiguity.

However if we instead have two different functions passed into `CombineTwo`:
```
operation Bar<'T> (arg : 'T) : Unit { }

operation Main() : Unit {
    (CombineTwo(Foo, Bar))("Hi", 3);
}
```
Now there is no more conflict because it sees `Foo.T` resolving to `String` and `Bar.T` resolving to `Int`. The fact that one of these scenarios is supported and not the other is the issue.

### LINKED ISSUE #662: Some non-concretized callables are allowed to be called or given as arguments
**Describe the bug**

The Q# language specification says in [Type Parameterizations](https://github.com/microsoft/qsharp-language/blob/main/Specifications/Language/4_TypeSystem/TypeParameterizations.md):

> A type parametrized callable needs to be concretized before it can be assigned or passed as argument, meaning all type parameters need to be resolved to concrete types.

The compiler doesn't implement this correctly, because the following programs compile:

```qsharp
function Foo<'a>() : Unit { }

@EntryPoint()
function Main() : Unit {
    Foo(); // 'a is not concrete.
}
```

```qsharp
function Undefined<'a>() : 'a {
    fail "Undefined";
}

function Foo<'a>(f : (Unit -> 'a)) : String {
    return "Foo";
}

@EntryPoint()
function Main() : Unit {
    Message(Foo(Undefined)); // 'a is not concrete.
}
```

**Expected behavior**

1. The compiler should forbid the programs above to comply with the language spec; or
2. Since a type system that allows these programs would still be sound, and it may actually be useful to allow this in certain cases, we could consider updating the language spec to loosen the restriction that all type parameters must be concretized.

For an example of why we should allow non-concrete types, consider this program that is not supported currently but would be reasonable to support:

```qsharp
@EntryPoint()
function Main() : Unit {
    let xs = [];
    Message($"{xs}");
}
```

The program should clearly print the string `"[]"`. But the type of `xs` cannot be more specific than `'a[]`, which is not concrete. In the case of empty arrays, the type variable `'a` is not used, though, so it does not actually need to be concrete.

**Additional context**

The example programs cause monomorphization to fail using PR #654:

> System.ArgumentException: Couldn't resolve all type parameters for App.Foo
>    at Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker.BuildCallGraph.ConcreteCallGraphWalker.<>c__DisplayClass2_0.<GetSpecializationFromRequest>b__1(QsLocalSymbol typeParam) in C:\Users\samarsha\Source\Repos\microsoft\qsharp-compiler\src\QsCompiler\Transformations\CallGraph\ConcreteCallGraphWalker.cs:line 101

This issue is based on this discussion: https://github.com/microsoft/qsharp-compiler/pull/597#discussion_r490666610

### LINKED ISSUE #857: QEP 2: Enhanced Array Literals
This issue tracks the implementation of [QEP 2: Enhanced Array Literals](https://github.com/microsoft/qsharp-language/blob/main/Approved/2-enhanced-array-literals.md). The feature branch is [feature/qep2](https://github.com/microsoft/qsharp-compiler/tree/feature/qep2).

- [x] Support the empty array constructor, `[]` (#884)
  - [x] Add Hindley-Milner type inference module (#917)
  - [x] Replace existing type inference with Hindley-Milner (#940)
  - [x] Update usages of obsolete range fields in `QsTypeParameter` and `UserDefinedType` (#952)
- [x] Support the repeated constant array constructor, `[x, size = n]` (#856)
  - [x] Support `[x, size = n]` in attribute arguments (#884)
- [x] Add `SizedArray` to NamespaceManager.fs (`ExpressionHash`) and QsharpCodeOutput.cs (#963)
- [x] Update optimizations for `SizedArray`: ConstantPropagation.fs, Evaluation.fs, HelperFunctions.fs (`isLiteral`) (#963)
- [x] Address issues from bug bash (#980)
- [ ] Update QIR generation for `SizedArray` expression kind

Deferred:

* Deprecate the `new Type[n]` array constructor (#956)

### PR DESCRIPTION
**Pull request by samarsha**
Closes #857. Closes #662. Closes #573.

Depends on microsoft/qsharp-runtime#643.

### CODE REVIEW

File: src/QsCompiler/DataStructures/Diagnostics.fs [hunk 1]
```diff
@@ -434,6 +463,11 @@ type DiagnosticItem =
         let ApplyArguments =
             DiagnosticItem.ApplyArguments args
             << function
+            | ErrorCode.TypeMismatch ->
+                "The type {1} does not match the type {0}.\nActual type:   {3}\nExpected type: {2}"
+            | ErrorCode.MissingBaseType ->
+                "The type {1} does not {0} the type {2}.\nLeft-hand type:  {3}\nRight-hand type: {4}"
```
Comment (bettinaheim) [inline src/QsCompiler/DataStructures/Diagnostics.fs:469]: What is the argument 0 here?

**Review by bettinaheim — APPROVED** [review]
Great update to the type inference! 
The latest improvements to the error messages make a big difference.

File: src/QsCompiler/DataStructures/Diagnostics.fs (hunk 1 above)
Comment (samarsha) [reply src/QsCompiler/DataStructures/Diagnostics.fs:469]: A description of the relationship between the two types. Possible values are "share a subtype with", "equal", or "share a base type with". I think maybe the error name `MissingBaseType` should be renamed to something like `MissingIntersection`.

Comment (samarsha) [reply src/QsCompiler/DataStructures/Diagnostics.fs:469]: Renamed to `TypeIntersectionMismatch`.

Comment (bettinaheim) [reply src/QsCompiler/DataStructures/Diagnostics.fs:469]: Thanks.

**Pull request merged**